### PR TITLE
Change rendering flow.

### LINKED ISSUE #48: Inline CLI Diff
Currently I'm developing inline CLI support for the diff package.

I'm pretty close to finalization, but in order to keep things structured, I need to make some severe code changes which might result in merge-conflicts for the master branch.

I expect this won't be a problem once the new code is reviewed and merge-conflicts can be resolved, but I strongly recommend to pause development of any of the files in directory `php-diff/lib/jblond/Diff/Renderer/` until the PR is merged.

### PR DESCRIPTION
**Pull request by DigiLive**
- Updated examples for cgi and cli output.
- Renamed RendererAbstract.php to MainRendererAbstract.php.
- Added default options to the abstract main Renderer.
- Change: Custom options which where set before wil not be lost when
  setting new options.
- Added a main renderer so custom (sub)renderers are easier to create.
- Added a sub-renderer interface.
- Merged HtmlArray.php into MainRenderer.php.
- Updated HTML renderers so they act as a sub renderer of the new main
  renderer.
- Updated existing text renderers so they make use of the renamed
  abstract renderer.
- Added Unified CLI text renderer which closes #48.
- Renamed HtmlArrayTest.php to MainRenderersTest.php.
- Updated MainRenderersTest to be compatible with the main renderer.
- Added a PHPUnit test and resource for InlineCli.
- Changed collapsed block from inline to block element for the HTML
  unified renderer.
- Update PHPUnit resource to reflect above change.
- Renamed PHPUnit resource of unified CLI text renderer to reflect other
  resource names.

### CODE REVIEW

File: example/cli.php [hunk 1]
```diff
@@ -2,18 +2,23 @@
...
+use jblond\Diff\Renderer\Text\InlineCli;
 use jblond\Diff\Renderer\Text\UnifiedCli;
 
-// Include and instantiate autoloader.
-require '../vendor/autoload.php';
+// Validate the interpreter.
+if (php_sapi_name() !== 'cli') {
+    echo 'This script demonstrates console support for the php-diff package.<br>';
+    echo 'Please execute it from a cli interpreter.';
+    exit;
```
Comment (JBlond) [inline example/cli.php:12]: Throw an exception instead of exit?

File: lib/jblond/Diff/Renderer/SubRenderer.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ * PHP version 7.2 or greater
+ *
+ * @package     jblond\Diff\Renderer\Html
+ * @author      Ferry Cools <info@DigiLive.nl>
+ * @copyright   (c) 2020 Ferry Cools
+ * @license     New BSD License http://www.opensource.org/licenses/bsd-license.php
+ * @version     2.0.0
+ * @link        https://github.com/JBlond/php-diff
+ */
+interface SubRenderer
```
Comment (JBlond) [inline lib/jblond/Diff/Renderer/SubRenderer.php:19]: Interfaces names should end with "Interface"

**Comment by JBlond** [comment]
I like the idea of the sub renderer class. I had something similar in my mind.

File: example/cli.php (hunk 1 above)
Comment (DigiLive) [reply example/cli.php]: Resolved at commit 0ef6def

File: lib/jblond/Diff/Renderer/SubRenderer.php (hunk 2 above)
Comment (DigiLive) [reply lib/jblond/Diff/Renderer/SubRenderer.php]: Resolved at commit 0ef6def

**Comment by DigiLive** [comment]
> 
> 
> I like the idea of the sub renderer class. I had something similar in my mind.

Great minds think alike. 😁
Using a main renderer (based on an abstract class) and sub renderers (based on an interface), hopefully makes it easier to create and maintain other renderers.

**Comment by JBlond** [comment]
Please resolve the merge conflict :)

**Comment by DigiLive** [comment]
My goal was to resolve the merge conflict after this branch is carefully tested as described above and approved by you.

Would you care about pulling the inlineCli branch and testing it locally?

**Comment by JBlond** [comment]
There is one issue with the inline unit test. 
"Trouble" part is
```PHP
 // This part of the line has been changed. Surround it with user defied markers.
                    $basePart    = $this->options['deleteMarkers'][0] . $part . $this->options['deleteMarkers'][1];
                    $changedPart =
                        $this->options['insertMarkers'][0] .
                        $changedLineParts[$partKey] .
                        $this->options['insertMarkers'][1];
```

That markers that you add are not part of the "diff" file. tests/resources/textInlineCli.txt

```bash
<title>Hello [0;30m[41m-World-[0m[0;30m[42m+You+[0m!</title>
```

While the output is 

```bash
<title>Hello -World-+You+!</title>
```


```diff
diff --git a/tests/resources/textInlineCli.txt b/tests/resources/textInlineCli.txt
index 3678f6f..64e1a2a 100644
--- a/tests/resources/textInlineCli.txt
+++ b/tests/resources/textInlineCli.txt
@@ -1,23 +1,26 @@
+-= Inline Marked =-
+
 =|<html>
 =|    <head>
 =|        <meta http-equiv="Content-type" content="text/html; charset=utf-8"/>
-x|        <title>Hello [0;30m[41m-World-[0m[0;30m[42m+You+[0m!</title>
+x|        <title>Hello -World-+You+!</title>
 =|    </head>
 =|    <body>
 =|        <h1>This is demo content to show features of the php-diff package.</h1>
--|[0;30m[41m        <h2>This line is removed from version2.</h2>[0m
+-|        <h2>This line is removed from version2.</h2>
+-|        <h2>This line is also removed from version2.</h2>
 =|        <h2>This line is the same for both versions.</h2>
-x|        <h2>This line has [0;30m[41m-inline -[0m[0;30m[42m++[0mdifferences between both versions.</h2>
+x|        <h2>This line has -inline -++differences between both versions.</h2>
 =|        <h2>This line is the same for both versions.</h2>
-x|        <h2>This line also has [0;30m[41m-inl-[0m[0;30m[42m+InL+[0mine differences between both versions.</h2>
+x|        <h2>This line also has -inl-+InL+ine differences between both versions.</h2>
 =|        <h2>This line is the same for both versions.</h2>
-+|[0;30m[42m        <h2>This line is added to version2.</h2>[0m
++|        <h2>This line is added to version2.</h2>
 =|
 =|        <p>
 =|            It's also compatible with multibyte characters (like Chinese and emoji) as shown below:
-x|            另外我覺得那個評[0;30m[41m-價-[0m[0;30m[42m+鑑+[0m的白色櫃子有點沒有必要欸。外觀我就不說了 ，怎麼連空間都那麼狹隘。不過倒是從這個地方看出所謂的“改革”
-x|            Do you know what "[0;30m[41m-金槍魚罐頭-[0m[0;30m[42m+魚の缶詰+[0m" means in Chinese?
-x|            [0;30m[41m-🍏🍎-[0m[0;30m[42m+🍎🍏+[0m🙂
+x|            另外我覺得那個評-價-+鑑+的白色櫃子有點沒有必要欸。外觀我就不說了 ，怎麼連空間都那麼狹隘。不過倒是從這個地方看出所謂的“改革”
+x|            Do you know what "-金槍魚罐頭-+魚の缶詰+" means in Chinese?
+x|            -🍏🍎-+🍎🍏+🙂
 =|        </p>
 =|
 =|        <p>Just some lines to demonstrate the collapsing of a block of lines which are the same in both versions.</p>
@@ -25,9 +28,9 @@ x|            [0;30m[41m-🍏🍎-[0m[0;30m[42m+🍎🍏+[0m🙂
 =|        <p>Just some lines to demonstrate the collapsing of a block of lines which are the same in both versions.</p>
 =|        <p>Just some lines to demonstrate the collapsing of a block of lines which are the same in both versions.</p>
 =|        <p>Just some lines to demonstrate the collapsing of a block of lines which are the same in both versions.</p>
-x|[0;30m[41m-        -[0m[0;30m[42m+        +[0m<h2>This line also has inline differences between both versions. It's the whitespace in front.</h2>
+x|-        -+        +<h2>This line also has inline differences between both versions. It's the whitespace in front.</h2>
 =|        <h2>This line is the same for both versions.</h2>
-x|        <h2>This line also has inline differences between both versions[0;30m[41m-.-[0m[0;30m[42m+!+[0m</h2>
+x|        <h2>This line also has inline differences between both versions-.-+!+</h2>
 =|    </body>
 =|</html>
 =|

```
on the output of the colored version I don't see the + and the - at all, but the colors

**Comment by DigiLive** [comment]
> That markers that you add are not part of the "diff" file. tests/resources/textInlineCli.txt
>
> ```shell
> <title>Hello �[0;30m�[41m-World-�[0m�[0;30m�[42m+You+�[0m!</title>
> ```
> 
> While the output is
> 
> ```shell
> <title>Hello -World-+You+!</title>
> ```

I'm not sure what you mean:
https://github.com/JBlond/php-diff/blob/0ef6def8a0c8b310a0b64a3a71da21e7f85c2974/tests/resources/textInlineCli.txt#L4
As you can see, the insertMarkers, deleteMarkers, equalityMarkers and escape sequences are included in the "diff" file.
Also, all PHPUnit tests pass.

Please note, the PHPUnit test sets different renderer options as the cli example script does.

- The example script outputs 1 inlineCli diff with all available markers and  1 inlineCli diff with none of the markers, but with colors (escape sequences).
- The PHPUnit test gets 1 inlineCli diff with available markers and colors (escape sequences).
  Like a merge of the two diff outputs at the bullet above. 

> on the output of the colored version I don't see the + and the - at all, but the colors

This is because I didn't define the insert- and deleteMarkers so they default to `''`.
Since the background of the changes is colored, the markers don't really have an added value.

I guess it's all about setting the options so they will meet expectations.
Once the PR is merged, I can update the read-me to include all available options.
Maybe it's a good idea to enable the Wiki of this repo.

---
Another issue might be the use of the `PHP_EOL` constant which might bite when line endings change when committing.
The line endings of the test "diff" files must match the line ending of the platform which runs the script.
I might replace `PHP_EOL` to `"\n"` to keep it cross-platform.

**Comment by JBlond** [comment]
As for the �characters. Do you have that as in your file, too? If not we might have to add a gitattributes file with type binary for that file and you have to recommit it.

**Comment by DigiLive** [comment]
> As for the �characters. Do you have that as in your file, too? If not we might have to add a gitattributes file with type binary for that file and you have to recommit it. 

Those characters are still in the file, but they're non printable characters.

In this case, they define the escape sequences (for coloring) which are now rendered as diamond question mark by the application you use for viewing the file content.

Viewing the file in notepad++ will represent them a different way.

The same goes for the emoji characters.
They're part of the utf8mb4 character set.
If your viewing application doesn't support that set, they show a replacement character like the diamond question mark or something.

**Comment by JBlond** [comment]
Thanks for the awesome work on this.

**Pull request merged**